### ISSUE TITLE
large genome settings for ABySS

### ISSUE THREAD

**Issue by rwhetten** [0x002A/MuCHSALSA#8]
The pipeline.sh script does not include a setting for the Bloom filter size, which is used during unitig assembly as I understand the ABySS assembler. For large genomes (multi-gigabase) the ABySS documentation suggests setting this value to reduce memory use relative to the MPI mode - would this present a problem for the rest of the MuCHSALSA pipeline? I'm working with a 22 Gb genome, so memory use is a concern.

**Comment by TGatter** [comment]
Sorry for the latish reply. 
Let me start with a short warning that we have not tested our pipeline for genomes of this massif size.
May I inquire what type of organism you are working on? Do you have already insights onto the duplication/repeat rates?

To answer your question I presume using bloom filter for abyss has a low influence on the assembly. It might reduce the length of unitigs to a certain degree, and thereby exhibit a slightly lower contiguity in the final MuCHSALSA assembly, but I presume this effect is small.

I have two  other notes about working on genomes of this size:
- Please make sure to use the latest development version, as it is drastically faster than previous versions on larger genomes. Unfortunately I have no ready made binary for it yet, but perhaps I can get someone to do it for me next week.
- You may optionally skip the scrubbing step in the pipeline to reduce overall runtimes. This step is rather expensive. This will leave heads and tails with very high error rates that are common for nanopore reads in the assembly, however.

I attached a version of the pipeline with a bloom setting and skipped scrubbing for you. Please be warned that I could not test it though. I currently can't sit for tool long and only have one hand to type unfortunately. Feel free to also contact me directly at thomas@bioinf.uni-leipzig.de for further discussions or results. I am very curious about your project and hope MuCHSALSA can yield good results also for you.

**Comment by rwhetten** [comment]
It would be great to try a compiled binary of the development version - I've tried installing clang and the other requirements for compiling, but haven't gotten it to work yet. I plotted a log-log version of the Jellyfish_histo output to see the distribution of kmers across frequency classes, and chose a frequency threshold that results in filtering out about 83% of the reads. This is consistent with published reports that the genome is about 82% repetitive sequences. The threshold I used (57) is considerably lower than that chosen by the setAbundanceThresholdFromHisto.py script (93) - I think the high repetitive-sequence content of this genome results in a very large IQR in the frequency distribution, so the script chooses a value that only excludes about 70% of the reads.

**Comment by rwhetten** [comment]
I was able to compile the development version on Ubuntu 20.04 after installing several required packages, and modifying the provided CMakeLists.txt file. For others interested in compiling, this is what worked for me. 
Install an executable binary of CMake v3.17.4 from https://github.com/Kitware/CMake/releases
Install compiler packages and development tools from Ubuntu repository, including checkinstall, clang-12, clang-tidy-12, libc++-12-dev, libc++abi1-12, libc++abi-12-dev, and libpthread-stubs0-dev
Install Doxygen, Graphviz, and associated packages from Ubuntu repository, including doxygen, libclang1-10, libllvm10, libxapian30, graphviz, libgraphviz-dev, libann0, libcdt5, libcgraph6, libgts-0.7-5, libgts-bin, libgvc6, libgvc6-plugins-gtk, libgvpr2, liblab-gamut1, libpathplan4, and libxdot4
Modify the MuCHSALSA/CMakeLists.txt file by including the following lines, following a suggestion at https://github.com/davisking/dlib/issues/2043 related to the CMake error `Could NOT find Threads (missing: Threads_FOUND)`:
`set(CMAKE_THREAD_LIBS_INIT "-lpthread")`
`set(CMAKE_HAVE_THREADS_LIBRARY 1)`
`set(CMAKE_USE_WIN32_THREADS_INIT 0)`
`set(CMAKE_USE_PTHREADS_INIT 1)`
`set(THREADS_PREFER_PTHREAD_FLAG ON)`
After installing those packages and editing the CMakeLists.txt file, the compiling process completed without error and `make test` returned a positive result.

**Comment by 0x002A** [comment]
First of all, I want to thank you for your feedback. I will have a look into it and check whether the files should be updated. During the initial development, MuCHSALSA compiled successfully on several unrelated machines running Ubuntu and having the listed prerequisites available. Of course, this might not hold for newer versions of cmake (since the problem seems to be that cmake cannot find pthreads).

Secondly, please make sure to compile MuCHSALSA in release mode since the performance differences between the development version and the release version are tremendous. This can be done by running cmake the following way:

`cmake -DCMAKE_BUILD_TYPE=Release ..`

Also I will try to provide a precompiled version based on the development branch within the next days.

We are eager to hear your feedback on MuCHSALSA.

Thanks again.

--
Kevin

**Comment by rwhetten** [comment]
My searches for information about the CMake failure to find pthreads.h pulled up posts from 2015 to the present, so it seems to be an ongoing issue that is not unique to a specific version of CMake. I'm a biologist, not a programmer, so I don't have more to offer regarding the nature of that problem. 
I used the CMake options `-DCMAKE_BUILD_TYPE=Release -DENABLE_BUILD_FAT_EXE=1` to produce a statically-linked release-version binary but haven't tested it yet on my data because I'm waiting on the assembler to finish assembling the "single-copy" subset of the genome. 
Regarding @TGatter's comment about the expense of the scrubbing step - for large genomes, using one of the accelerated versions of minimap2, either [mm2-fast](https://doi.org/10.1101/2021.07.21.453294) or [mm2-ax](https://www.biorxiv.org/content/10.1101/2022.03.09.483575v2.full.pdf), and more threads could speed up the mapping phase. The scrubber_bfs.py script probably needs to see all the data in order to bring the assembly together, and I can imagine that will take some time. Is it possible to make that step multi-threaded?

**Comment by 0x002A** [comment]
@rwhetten First of all, I apologise for the long response time, I could reproduce the error you had with the pthreads check on a fresh Ubuntu installation. In my experience, the problem resulted from the fact that cmake internally compiles a small test program to check whether pthreads is available and this compilations fails for a reason that is **not** related to pthreads being unavailable. In my case, it was libunwind-dev which was missing in order for cmake to successfully compile its test program and report pthreads as being available. I recommend to install the whole LLVM toolchain via their apt repository as described [here](https://apt.llvm.org) Additionally, you can look at the .github/workflows/build.yml file which describes the exact steps required to prepare a fresh Ubuntu based machine and build the MuCHSALSA Nightly Build. Whenever we merge a pull request, the Nightly Build is updated and you can download the binary and benefit from the latest changes. You can find the Nightly Build [here](https://github.com/0x002A/MuCHSALSA/releases/tag/nightly). Furthermore, I updated the README so that it includes a more complete list of packages that need to be installed for a successful build. As the technical problems seem to be resolved I am going to close this issue. Feel free to open a new issue if you experience new technical problems.
